database/interface.go

```diff
@@ -271,18 +286,22 @@ func (i *Interface) PutNew(r record.Record) (err error) {
 	}
 
 	r.Lock()
-	defer r.Unlock()
-
 	if r.Meta() != nil {
 		r.Meta().Reset()
 	}
 	i.options.Apply(r)
+	remove := r.Meta().IsDeleted()
+	ttl := r.Meta().GetRelativeExpiry()
+	r.Unlock()
 
-	written := i.updateCache(r, true)
+	// The record may not be locked when updating the cache.
+	written := i.updateCache(r, true, remove, ttl)
 	if written {
```

Comment:
Hm, but the cache uses `r.Key()` which needs to be guarded by the lock as a concurrent `SetKey()`. I know this is rather unlikely as there must have been a key set already but in theory we have an issue here.

edit: also, it's not clear to me why the record must not be locked when updating the cache. May add this requirement to the `updateCache` function comment.